### PR TITLE
MINIFICPP-1784 MINIFICPP-1780 Agent restart after C2 property update

### PR DESCRIPTION
**Pull request by szaszm**
Changes to the extension path don't apply yet, because I did not yet get rid of the singleton `ExtensionManager`. I won't do that in the context of this PR.

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/src/c2/C2Agent.cpp [hunk 1]
```diff
@@ -803,6 +789,7 @@ utils::TaskRescheduleInfo C2Agent::consume() {
       extractPayload(C2Payload{ Operation::HEARTBEAT });
     }
   }
+  if (restart_needed_) { request_restart_(); }
```
Comment (szaszm) [inline libminifi/src/c2/C2Agent.cpp:792]: The restart happens here, after consuming all of the C2 commands.

**Review by lordgamez — APPROVED** [review]
LGTM, just one minor comment about naming.

File: libminifi/include/utils/WorkerThread.h [hunk 2]
```diff
@@ -49,8 +49,9 @@ class WorkerThread final {
 
 /**
  * A worker that executes arbitrary functions with no parameters asynchronously on an internal thread, returning a future to the result.
+ * This is different from the Worker in ThreadPool.h in that it's only capable of running tasks, and any scheduling is left for the user.
  */
-class Worker final {
+class Worker2 final {
```
[the remark this answers is not in the document]
Comment (martinzink) [reply libminifi/include/utils/WorkerThread.h:54]: I was just about to comment the same thing.
Based on the comment above it, we could name it ManuallyScheduledWorker or something like that?

[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/utils/WorkerThread.h:54]: Sounds good

[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/utils/WorkerThread.h]: [51eac31](https://github.com/apache/nifi-minifi-cpp/pull/1299/commits/51eac3191d6be532b80c51e41ce664edc88e0d3e)

**Review by martinzink — APPROVED** [review]

File: libminifi/include/utils/meta/type_list.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+template<typename... Types>
+struct type_list {
+  template<typename T>
+  [[nodiscard]] constexpr static bool contains() noexcept {
+    return (std::is_same_v<T, Types> || ...);
+  }
+
+  template<template<typename...> typename Metafn>
+  using apply = Metafn<Types...>;
```
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/utils/meta/type_list.h:31]: Calls a metafunction with the types as arguments. It's not used at the moment, I just thought that it's an essential feature that might come handy later.

File: libminifi/include/core/state/Value.h [hunk 4]
```diff
@@ -497,34 +495,31 @@ static inline std::shared_ptr<Value> createValue(const double &object) {
...
-  ValueNode()
-      : value_(nullptr) {
-  }
+  using supported_types = utils::meta::type_list<int, uint32_t, size_t, int64_t, uint64_t, bool, char*, const char*, double, std::string>;
 
+ public:
+  ValueNode() = default;
   ValueNode(ValueNode &&vn) = default;
   ValueNode(const ValueNode &vn) = default;
+  ValueNode &operator=(const ValueNode &ref) = default;
```
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/core/state/Value.h:504]: An oversight, will fix.

File: libminifi/include/utils/meta/type_list.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgerlits) [reply libminifi/include/utils/meta/type_list.h:31]: I don't insist, but I would remove it if it is not needed at the moment.

File: libminifi/src/c2/C2Client.cpp [hunk 5]
```diff
@@ -68,7 +71,23 @@ void C2Client::initialize(core::controller::ControllerServiceProvider *controlle
     logger_->log_info("Agent class is not predefined");
   }
 
-  configuration_->setFallbackAgentIdentifier(getControllerUUID().to_string());
+  // Set a persistent fallback agent id. This is needed so that the C2 server can identify the same agent after a restart, even if nifi.c2.agent.identifier is not specified.
```
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/c2/C2Client.cpp:74]: I found it complicated to change the identifier/fallback identifier logic, but I can introduce a new property called nifi.c2.agent.identifier.fallback, and write to that. What do you think?

**Review by fgerlits — APPROVED** [review]

**Pull request closed without merging**